cmd/snap-repair/runner.go

```diff
@@ -185,3 +211,268 @@ func (run *Runner) Peek(brandID, repairID string) (headers map[string]interface{
+	d := filepath.Join(dirs.SnapRepairAssertsDir, brandID, repairID)
+	err := os.MkdirAll(d, 0775)
+	if err != nil {
+		return err
+	}
+	buf := &bytes.Buffer{}
+	enc := asserts.NewEncoder(buf)
+	for _, a := range r {
+		if err := enc.Encode(a); err != nil {
+			return fmt.Errorf("cannot encode repair assertions %q-%s for saving: %v", brandID, repairID, err)
+		}
+	}
+	p := filepath.Join(d, fmt.Sprintf("repair.r%d", r[0].Revision()))
+	return osutil.AtomicWriteFile(p, buf.Bytes(), 0600, 0)
+}
+
+func (run *Runner) readSavedStream(brandID string, seq, revision int) (r []asserts.Assertion, err error) {
+	repairID := strconv.Itoa(seq)
+	d := filepath.Join(dirs.SnapRepairAssertsDir, brandID, repairID)
+	p := filepath.Join(d, fmt.Sprintf("repair.r%d", revision))
+	f, err := os.Open(p)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	dec := asserts.NewDecoder(f)
+	for {
+		a, err := dec.Decode()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("cannot decode repair assertions %q-%s from disk: %v", brandID, repairID, err)
+		}
+		r = append(r, a)
+	}
+	if err := checkStreamSanity(brandID, repairID, r); err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+func (run *Runner) makeReady(brandID string, idx int) (state *RepairState, repair *asserts.Repair, err error) {
+	sequence := run.state.Sequences[brandID]
+	var a []asserts.Assertion
+	if idx < len(sequence) {
+		// consider retries
+		state = sequence[idx]
+		if state.Status != RetryStatus {
+			return nil, nil, errNotToRun
+		}
+		var err error
+		a, err = run.refetch(brandID, state.Seq, state.Revision)
+		if err != nil {
+			// TODO: log error
+			// try to use what we have already on disk
+			a, err = run.readSavedStream(brandID, state.Seq, state.Revision)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+	} else {
+		// fetch the next repair in the sequence
+		nextSeq := 0
+		if n := len(sequence); n != 0 {
+			nextSeq = sequence[n-1].Seq
+		}
+		nextSeq += 1
+		var err error
+		a, err = run.fetch(brandID, nextSeq)
+		if err != nil && err != errNotToRun {
+			return nil, nil, err
+		}
+		state = &RepairState{
+			Seq: nextSeq,
+		}
+		if err == errNotToRun {
+			// TODO: store headers to justify decision
+			state.Status = NotApplicableStatus
+			return state, nil, errNotToRun
+		}
+	}
+	// TODO: verify with signatures
+	if err := run.saveStream(brandID, state.Seq, a); err != nil {
+		return nil, nil, err
+	}
+	repair = a[0].(*asserts.Repair)
+	state.Revision = repair.Revision()
+	if !run.Applicable(repair.Headers()) {
+		state.Status = NotApplicableStatus
+		return state, nil, errNotToRun
+	}
+	return state, repair, nil
+}
+
+// Next returns the next repair for the brand id sequence to run/retry or ErrRepairNotFound if there is none atm. It updates the state as required.
+func (run *Runner) Next(brandID string) (*asserts.Repair, error) {
+	nextIdx, _ := run.nextIdx[brandID]
+	for {
+		state, repair, err := run.makeReady(brandID, nextIdx)
```

Comment:
I found the mixed use of index and sequence in the code somewhat hard to read at times, and this may end up creating errors due to crossing of context and the difference in indexing (zero vs. one based).

I suggest normalizing around sequences everywhere except for traditional loops in which we'd use "i" anyway.